### PR TITLE
envoy_build_system: refactor main bzl file into separate smaller files

### PR DESCRIPTION
**Pull request by fcfort**
Description: This PR refactors `envoy_build_system.bzl` into 6 separate bzl files with the following dependencies:
envoy_build_system_refactor

No changes were made to the existing build macros themselves, only to their location.

The aim is to simplify this file and aid understanding and modification of the build system in the future.

Each of the separate files takes some of the build macros from `envoy_build_system.bzl`, e.g. `envoy_binary.bzl` has the macros for creating binaries, `envoy_test.bzl` has the macros for creating test targets and so on.

It was necessary to have `envoy_build_system.bzl` depend upon all other bzl files in order to not break backwards compatibility with `load()` statements from `BUILD` files that referred to `envoy_build_system.bzl`.

It was also necessary to move the public `envoy_select_*` macros from `envoy_build_system.bzl` into  `envoy_select.bzl` in order to avoid a circular dependency.

All macros in `envoy_internal.bzl` are meant to be private, but bazel has no concept of package private macros, so this was the best I could do. I could rename it `_envoy_internal` or move it to a sub-directory if we really want to hide it. I have taken to `load()`ing macros from this file using an underscore, e.g.  
  ```
  load(":envoy_internal.bzl", _foo = "foo")  
  ```

If going forward we wish to treat `envoy_build_system.bzl` as the only bzl file one should ever `load()` from, then this PR is sufficient to accomplish that. If we want developers to import the separate files (e.g. `envoy_binary.bzl`, `envoy_test.bzl`), I can work to migrate all existing uses within Envoy to the new files and mark the symbols referenced from `envoy_build_system.bzl` as deprecated (perhaps via `print` statement warnings).

Risk Level: Medium
Testing: Ran `bazel build //...`
Docs Changes: None necessary, this PR maintains backwards compatibility with existing `load()`s of `envoy_build_system.bzl`.
Release Notes: N/A

### CODE REVIEW

File: bazel/envoy_build_system.bzl [hunk 1]
```diff
@@ -652,54 +150,41 @@ def envoy_proto_descriptor(name, out, srcs = [], external_deps = []):
...
-    return select({
-        "@envoy//bazel:force_libcpp": if_libcpp,
-        "@envoy//bazel:apple": [],
-        "@envoy//bazel:windows_x86_64": [],
-        "//conditions:default": default or [],
-    })
+# Here we create wrappers for each of the public targets within the separate bazel
+# files loaded above. This is necessary so that consumers who had previously used
+# macros imported from envoy_build_system.bzl don't have to change their BUILD
+# files to reference the new location.
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: I think it's fine to keep this as the permanent abstraction layer, i.e. a single import of `envoy_build_system.bzl`. The rationale is that we don't want to create too much reliance on the internal structure of the `.bzl` file elsewhere, and the surface of exported macros is small enough that we don't need to have fine grained `.bzl` disaggregation. If that sounds reasonable, maybe update the comments here to reflect.

File: bazel/envoy_build_system.bzl [hunk 2]
```diff
@@ -1,7 +1,30 @@
...
+    _envoy_proto_library = "envoy_proto_library",
+)
+load(
+    ":envoy_select.bzl",
+    _envoy_select_google_grpc = "envoy_select_google_grpc",
+    _envoy_select_hot_restart = "envoy_select_hot_restart",
+)
+load(
+    ":envoy_test.bzl",
+    _envoy_cc_fuzz_test = "envoy_cc_fuzz_test",
```
Comment (htuch) [inline bazel/envoy_build_system.bzl]: So, is the issue that if you import `envoy_cc_fuzz_test` via `load`, it doesn't immediately give you a re-export?

Comment (fcfort) [reply bazel/envoy_build_system.bzl]: Yeah, exactly. Simply `load`ing it doesn't export the symbol. You have to declare it in the bzl file to export it. Even doing `load(foo = "foo")` doesn't export `foo`.

File: bazel/envoy_build_system.bzl (hunk 1 above)
Comment (fcfort) [reply bazel/envoy_build_system.bzl]: Sounds good, done. I've updated the comments throughout the files to instruct readers to only `load()` `envoy_build_system.bzl`. We can also prepend envoy_library.bzl, envoy_binary.bzl, etc. with an `_` to further discourage their use.

**Pull request merged**